File: scala/src/test/scala/bbc/dojonorth/lyrics/LoadLyricsDataSpec.scala
package bbc.dojonorth.lyrics

import org.scalatest.{FlatSpec, Matchers}

import scala.collection.immutable.Nil

class LoadLyricsDataSpec extends FlatSpec with Matchers {

  val fixturesDir = "./src/test/resources"
  val testFile = s"$fixturesDir/lyricdata_sample.txt"

  "A lyric data loader" should "ignore comment lines" in {
    val lines = LoadLyricData.loadDataFile(testFile)
    val commentedLines = lines filter { line => line.startsWith("#") } toList

    commentedLines should be (Nil)
  }

  it should "contain the words list in the first element" in {
    val lines = LoadLyricData.loadDataFile(testFile).toList
    val firstElement: String = lines.head

    firstElement should startWith ("%")
    firstElement should include ("the")
    firstElement should include ("love")
  }

  it should "contain a track sample in the second element" in {
    val lines = LoadLyricData.loadDataFile(testFile).toList

    val secondElement = lines.tail.head // wish I had Scheme's cadr here!

    secondElement should startWith ("TRA")
    secondElement should include ("1548880")
  }

  val wordsLine = "%i,the,you,to,and,a,Me,it,not,in,my,is,of,your,"

  "A words line parser" should "produce words from the string" in {
    val parsedLine = LoadLyricData.parseWordsLine(wordsLine)

    parsedLine should not contain ("%i")
    parsedLine should contain ("i")
    parsedLine should contain inOrder ("of","your")
    parsedLine should have size 14
  }
  it should "only produce lowercase words" in {
    val parsedLine = LoadLyricData.parseWordsLine(wordsLine)

    parsedLine foreach {
      w =>
        w.toLowerCase should be (w)
    }
  }

  "A track line parser" should "extract the track ID, the MusixMatch ID and a map of word ID to count" in {
    val lines = LoadLyricData.loadDataFile(testFile).toList
    val words = LoadLyricData.parseWordsLine(lines.head)

    // helper to make sure 1-based indexing is adhered to
    def getWordForIndex(index: Int) = words(index-1)

    val parsedTrack = LoadLyricData.parseTrackLine(lines.tail.head, words)

    parsedTrack.trackId should be ("TRAABRX12903CC4816")
    parsedTrack.mxmId should be (1548880)
    parsedTrack.words should contain (getWordForIndex(2) -> 19)
    parsedTrack.words should contain (getWordForIndex(30) -> 11)

  }

  val dummySong: SongWords = SongWords("TRAAMYTEST", 1000, Map("the" -> 19, "to" -> 7, "and" -> 6))

  "A song parser" should "produce a stream of 2 songs from the test file" in {

    val songs = LoadLyricData.load(testFile).songs.toList

    songs should contain (dummySong)
    songs.length should be (2)
  }

  "A lyric data loader" should "produce a valid lyric data object" in {
    val fullyParsedLyricData = LoadLyricData.load(testFile)

    fullyParsedLyricData.words should contain ("this")
    fullyParsedLyricData.words should not contain ("jens")


    val songs = fullyParsedLyricData.songs.toList
    songs should contain (dummySong)
    songs.length should be (2)
  }

}

File: javascript/test/lyrics/LoadLyricsDataSpec.js
var assert = require('assert');
var _ = require('lodash');
var LyricDataLoader = require('../../lib/lyrics/LyricDataLoader');

describe('LyricDataLoader', function () {
    var lyricDataLoader, fixturesDir, testFile;
    before(function () {
        lyricDataLoader = new LyricDataLoader();
        fixturesDir = './test/resources';
        testFile = fixturesDir + '/lyricdata_sample.txt';
    });

    describe('#loadDataFile()', function () {

        it('should ignore comment lines', function () {
            var lines = lyricDataLoader.loadDataFile(testFile);
            var commentedLines = _.filter(lines, function(line) {
                return line.indexOf('#') === 0;
            });

            assert.equal(commentedLines.length, 0);
        });

        it('should contain the wordlist in first element', function () {
            var lines = lyricDataLoader.loadDataFile(testFile);
            var firstElement = lines[0];

            assert.equal(0, firstElement.indexOf('%'));
            assert.notEqual(-1, firstElement.indexOf('the'));
            assert.notEqual(-1, firstElement.indexOf('love'));
        });

        it('should contain a track sample in second element', function () {
            var lines = lyricDataLoader.loadDataFile(testFile);
            var secondElement = lines[1];

            assert.equal(0, secondElement.indexOf('TRA'));
            assert.notEqual(-1, secondElement.indexOf('1548880'));
        });
    });

    describe('#parseWordsLine()', function () {
        var wordsLine = "%i,the,you,to,and,a,Me,it,not,in,my,is,of,your,";

        it('should produce words from the string', function () {
            var parsedLine = lyricDataLoader.parseWordsLine(wordsLine);

            assert.equal(-1, parsedLine.indexOf('%i'));
            assert.notEqual(-1, parsedLine.indexOf('i'));
            assert(parsedLine.indexOf('of') < parsedLine.indexOf('your'));
            assert.equal(14, parsedLine.length);
        });

        it('should produce lowercase only', function () {
            var parsedLine = lyricDataLoader.parseWordsLine(wordsLine);

            _.each(parsedLine, function(word) {
                assert.equal(word, word.toLowerCase());
            });

        });
    });

    describe('#parseTrackLine()', function () {
        it('should extract the track ID, the MusixMatch ID and a map of word ID to count', function () {
            var lines = lyricDataLoader.loadDataFile(testFile);
            var words = lyricDataLoader.parseWordsLine(lines[0]);

            var getWordForIndex = function (n) {
                return words[n-1];
            };

            var parsedTrack = lyricDataLoader.parseTrackLine(lines[1], words);

            assert.equal('TRAABRX12903CC4816', parsedTrack.trackId);
            assert.equal('1548880', parsedTrack.mxmId);
            assert.equal(19, parsedTrack.wordMap[getWordForIndex(2)]);
            assert.equal(11, parsedTrack.wordMap[getWordForIndex(30)]);
        });
    });

    describe('#load', function () {
        it('should produce 2 songs from the test file', function () {
            var dummySong = {
                trackId: 'TRAAMYTEST',
                mxmId: 1000,
                wordMap: {
                    'the': 19,
                    'to': 7,
                    'and': 6
                }
            };

            var songs = lyricDataLoader.load(testFile).songs

            assert.equal(2, songs.length);
            assert.deepEqual(dummySong, songs[1]);
        });
    });
});
